File: xstudioz/feeds.py
"""Every feed this system runs on, and whether it is actually current.

WHY THIS MODULE EXISTS

The system has six inputs. Each one was individually checkable and none of them
was checked together, so the honest answer to "is the system live?" was "open
six things and work it out". Nobody does that daily, which means the real answer
was always "probably", and on 2026-08-07 "probably" was wrong by a whole
verdict: the engine on a 44-hour-old snapshot reported organic in BREACH at
index 26, and the same engine on the live snapshot reported HEALTHY at 79.

A feed has three states and they are not the same thing:

    live        it answered, and its newest data is recent enough to use
    stale       it answered, and its newest data is older than it should be
    unreachable it did not answer, or was never configured

`stale` and `unreachable` are both bad and they need different fixes: stale
means somebody has stopped filling something in, unreachable means a credential
or a service. Collapsing them into "not ok" sends whoever is on it to the wrong
place, so they stay distinct all the way to the screen.

WHAT THIS MODULE DOES NOT DO

It does not probe the two Supabase stores. Their keys belong to the hub, not to
the engine, and copying a credential into a second place so a status page can be
tidy is how credentials get committed. The hub probes those itself and joins
them to this on the page. Each side checks what it already has the right to
read.
"""
from __future__ import annotations

import datetime as _dt
import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

#: Hours after which a feed is stale. These are not uniform, because the feeds
#: do not update on the same clock and one threshold would either cry wolf on
#: the slow ones or stay silent on the fast ones.
THRESHOLDS_H: dict[str, float] = {
    # Fiverr publishes a day's figures around midday the next day, so 26 hours
    # is the normal maximum for a healthy morning and anything past it is a gap.
    "sheets": 26.0,
    # The gig page is scraped by hand because Fiverr blocks a plain fetch. A
    # week is tolerable for level and review counts; past that the rating maths
    # is describing a profile that has moved.
    "gig": 168.0,
    # The CSR handoff tracker is uploaded, not fetched. Two weeks.
    "tracker": 336.0,
}


@dataclass
class Feed:
    key: str
    label: str
    #: Where it comes from, in words a person can act on.
    source: str
    status: str = "unreachable"          # live | stale | unreachable
    as_of: str | None = None             # ISO date or datetime of newest data
    age_hours: float | None = None
    detail: str = ""
    #: What to do about it, when it is not live. Empty when it is.
    fix: str = ""

    def as_dict(self) -> dict[str, Any]:
        return {
            "key": self.key, "label": self.label, "source": self.source,
            "status": self.status, "as_of": self.as_of,
            "age_hours": self.age_hours, "detail": self.detail, "fix": self.fix,
        }


def _age_hours(when: _dt.datetime, now: _dt.datetime) -> float:
    if when.tzinfo is None:
        when = when.replace(tzinfo=_dt.timezone.utc)
    return (now - when).total_seconds() / 3600.0


def _classify(feed: Feed, threshold_h: float) -> None:
    if feed.age_hours is None:
        feed.status = "unreachable"
        return
    feed.status = "live" if feed.age_hours <= threshold_h else "stale"


def sheets_feed(intake: dict[str, Any] | None, now: _dt.datetime) -> Feed:
    """The order and inquiry workbooks, via the Apps Script snapshot."""
    f = Feed(
        key="sheets",
        label="Order and inquiry workbooks",
        source="Apps Script snapshot endpoint",
    )
    intake = intake or {}
    path = intake.get("path")
    f.age_hours = intake.get("age_hours")
    f.as_of = intake.get("generated_at")

    if path == "markdown":
        f.status = "unreachable"
        f.detail = ("built from a markdown export, which cannot separate profiles "
                    "and truncates")
        f.fix = "Set XSTUDIOZ_SNAPSHOT_URL and XSTUDIOZ_SNAPSHOT_TOKEN and re-run."
        return f

    _classify(f, THRESHOLDS_H["sheets"])
    if path == "live":
        f.detail = f"fetched live, {intake.get('tables', '?')} tables"
    elif path == "disk":
        f.detail = "read from a snapshot on disk; the live endpoint was " + (
            "not reachable" if intake.get("endpoint_configured") else "not configured")
        if not intake.get("endpoint_configured"):
            f.fix = ("Export XSTUDIOZ_SNAPSHOT_URL and XSTUDIOZ_SNAPSHOT_TOKEN "
                     "before the run. Without them every run silently uses "
                     "whatever snapshot happens to be on disk.")
        else:
            f.fix = "The endpoint is configured but did not answer. Check the Apps Script deployment."
    return f


def impressions_source_feed(intake: dict[str, Any] | None, now: _dt.datetime) -> Feed:
    """Which source today's impressions actually came from.

    Separate from the freshness row below, because "the numbers are old" and
    "the engine is reading the wrong place" are different sentences with
    different fixes, and for nine days in August the second one was the true
    one while only the first was on the page.

    The sheet stopped on 2026-08-06 when the team moved to the board. Reading
    the sheet after that is not a stale feed, it is the wrong feed, and it will
    never come right no matter who is asked to fill something in.
    """
    f = Feed(key="impressions_source", label="Impressions source",
             source="impressions board, via IMPRESSIONS_SUPABASE_KEY")
    note = ((intake or {}).get("impressions") or {})
    if note.get("source") == "board":
        f.status = "live"
        f.age_hours = 0.0
        f.as_of = note.get("board_newest")
        f.detail = (f"{note.get('board_rows', '?')} row(s) from the board, "
                    f"overriding {note.get('sheet_rows', '?')} from the sheet")
        return f
    f.status = "unreachable"
    f.detail = "impressions came from the sheet, which stopped on 2026-08-06"
    f.fix = ("Set IMPRESSIONS_SUPABASE_KEY where the engine runs. It is the "
             "anon key of the project the hub already reads. Without it the "
             "engine reads a sheet nobody fills in any more.")
    return f


def impressions_feed(root: Path, now: _dt.datetime, profile: str = "X Studioz",
                     records: list[Any] | None = None) -> Feed:
    """The Daily Data Sheet, via the same snapshot as the workbooks.

    Its own row rather than folded into `sheets`, because it is a different tab
    filled in by a different person on a different rhythm. It was 10 days
    behind while the order workbook was current, and one combined row would
    have reported the fresher of the two and hidden that entirely.

    NOTE THIS IS THE ENGINE'S COPY, NOT WHAT /entry SHOWS. From 2026-08-07 the
    hub reads XStudioz impressions from the impressions board, on Ezan's
    instruction; the board is fed from this same sheet and holds the corrected
    figures. The engine still ingests the sheet for its reach-versus-conversion
    diagnosis, so both rows appear on /feeds and a gap between them is a real
    signal rather than a duplicate.
    """
    f = Feed(key="impressions_sheet", label="Impressions (engine copy)",
             source="board, merged over the Daily Data Sheet")

    # IN-MEMORY FIRST, AND WHY IT MATTERS.
    #
    # This read the normalized file off disk, and `collect()` runs BEFORE
    # `write_normalized`, so it was reporting the PREVIOUS run's impressions
    # every morning -- always exactly one run behind. Harmless while the sheet
    # moved a day at a time and invisible for the same reason. It stopped being
    # harmless the morning the board arrived nine days fresher than the sheet:
    # the row went on saying 2026-08-06 while the file it names had just been
    # rebuilt to 2026-08-15.
    #
    # Passing the records the run actually computed removes the ordering
    # dependency rather than reordering two calls and hoping nobody swaps them
    # back. Disk stays as the fallback for callers that have no run in hand.
    if records is not None:
        newest_day = None
        for rec in records:
            if str(getattr(rec, "profile", "") or "") != profile:
                continue
            day = getattr(rec, "date", None)
            day = day.isoformat() if hasattr(day, "isoformat") else str(day or "")[:10]
            if day and (newest_day is None or day > newest_day):
                newest_day = day
        if not newest_day:
            f.detail = f"no row for {profile} in this run"
            f.fix = "The board and the sheet both hold nothing for this profile."
            return f
        f.as_of = newest_day
        when = _dt.datetime.fromisoformat(f"{newest_day}T12:00:00+00:00")
        f.age_hours = _age_hours(when, now)
        _classify(f, THRESHOLDS_H["sheets"] + 24)
        f.detail = f"newest row {newest_day}"
        if f.status != "live":
            f.fix = (f"Nothing has been entered on the impressions board since "
                     f"{newest_day}. The sheet behind it stopped on 2026-08-06 "
                     f"and is not coming back.")
        return f

    path = root / "data" / "normalized" / "impressions.jsonl"
    if not path.exists():
        f.detail = "the last run produced no impressions file"
        f.fix = "Run the engine. If it runs and this stays empty, the sheet tab was refused."
        return f
    newest = None
    try:
        import json as _json
        for line in path.read_text(encoding="utf-8").splitlines():
            if not line.strip():
                continue
            rec = _json.loads(line)
            if str(rec.get("profile") or "") != profile:
                continue
            day = str(rec.get("date") or "")[:10]
            if day and (newest is None or day > newest):
                newest = day
    except (OSError, ValueError) as exc:
        f.detail = f"impressions.jsonl could not be read: {exc}"
        return f

    if not newest:
        f.detail = f"no row for {profile}"
        f.fix = "The sheet holds no day for this profile, or the tab was refused."
        return f
    f.as_of = newest
    when = _dt.datetime.fromisoformat(f"{newest}T12:00:00+00:00")
    f.age_hours = _age_hours(when, now)
    _classify(f, THRESHOLDS_H["sheets"] + 24)  # filled in a day behind, like the rest
    f.detail = f"newest row {newest}"
    if f.status != "live":
        f.fix = f"Nobody has filled in the Daily Data Sheet since {newest}."
    return f


def gig_feed(root: Path, now: _dt.datetime) -> Feed:
    """The live Fiverr gig page capture."""
    f = Feed(
        key="gig",
        label="Fiverr gig page",
        source="browser capture into data/raw/gig/",
    )
    d = root / "data" / "raw" / "gig"
    files = sorted(d.glob("*.json")) if d.exists() else []
    if not files:
        f.detail = "no capture has ever been taken"
        f.fix = ("Fiverr returns 403 to a plain fetch. Use a browser-rendering "
                 "fetch tool on the gig URL in config/sources.yml and write the "
                 "result to data/raw/gig/<date>.json.")
        return f

    newest = files[-1]
    try:
        day = _dt.date.fromisoformat(newest.stem)
        when = _dt.datetime.combine(day, _dt.time(12, 0), tzinfo=_dt.timezone.utc)
    except ValueError:
        f.detail = f"{newest.name} is not named as a date"
        return f

    f.as_of = day.isoformat()
    f.age_hours = _age_hours(when, now)
    _classify(f, THRESHOLDS_H["gig"])
    f.detail = f"captured {day.isoformat()}"
    if f.status != "live":
        f.fix = ("Re-capture the gig page. It carries level, review total, the "
                 "star histogram, queue depth and response time, and the rating "
                 "maths is built on the histogram.")
    return f


def tracker_feed(root: Path, now: _dt.datetime) -> Feed:
    """The CSR handoff tracker workbook, uploaded by hand."""
    f = Feed(
        key="tracker",
        label="CSR handoff tracker",
        source="uploaded into data/raw/order_tracker/",
    )
    d = root / "data" / "raw" / "order_tracker"
    files = sorted(d.glob("*.xlsx"), key=lambda p: p.stat().st_mtime) if d.exists() else []
    if not files:
        f.detail = "nothing uploaded"
        f.fix = "Upload the tracker export into data/raw/order_tracker/."
        return f
    newest = files[-1]
    when = _dt.datetime.fromtimestamp(newest.stat().st_mtime, tz=_dt.timezone.utc)
    f.as_of = when.date().isoformat()
    f.age_hours = _age_hours(when, now)
    _classify(f, THRESHOLDS_H["tracker"])
    f.detail = newest.name
    if f.status != "live":
        f.fix = "Upload a fresh tracker export."
    return f


def snapshot_endpoint_feed(now: _dt.datetime) -> Feed:
    """Whether the credentials for the sheet feed are even present.

    Separate from `sheets_feed` on purpose. "The endpoint is not configured in
    this shell" and "the data is old" are different sentences with different
    fixes, and the second one is the symptom of the first.
    """
    f = Feed(
        key="snapshot_endpoint",
        label="Snapshot credentials",
        source="XSTUDIOZ_SNAPSHOT_URL / _TOKEN in the environment",
    )
    if os.environ.get("XSTUDIOZ_SNAPSHOT_URL") and os.environ.get("XSTUDIOZ_SNAPSHOT_TOKEN"):
        f.status = "live"
        f.age_hours = 0.0
        f.detail = "both present in this shell"
    else:
        f.status = "unreachable"
        f.detail = "not set in this shell, so the run cannot fetch live"
        f.fix = ("They live in the daily Routine's prompt. Any run started by "
                 "hand must export them first or it will quietly use whatever "
                 "snapshot is already on disk.")
    return f


def collect(root: Path, intake: dict[str, Any] | None = None,
            now: _dt.datetime | None = None,
            impressions: list[Any] | None = None) -> list[dict[str, Any]]:
    """Every feed the engine can check for itself."""
    now = now or _dt.datetime.now(_dt.timezone.utc)
    feeds = [
        snapshot_endpoint_feed(now),
        sheets_feed(intake, now),
        impressions_source_feed(intake, now),
        impressions_feed(root, now, records=impressions),
        gig_feed(root, now),
        tracker_feed(root, now),
    ]
    return [f.as_dict() for f in feeds]


def worst(feeds: list[dict[str, Any]]) -> str:
    """The overall verdict. Unreachable beats stale beats live: the page leads
    with the worst thing, because leading with the best is how a broken feed
    stays broken."""
    states = {f.get("status") for f in feeds}
    if "unreachable" in states:
        return "unreachable"
    if "stale" in states:
        return "stale"
    return "live"

File: system/views/feeds.js
// views/feeds.js, every input this business runs on, and whether it is current.
//
// WHY THIS PAGE EXISTS
//
//   Six feeds. Each individually checkable, none checked together, so the
//   honest answer to "is the system live?" was "open six things and work it
//   out". Nobody does that daily, which means the real answer was always
//   "probably".
//
//   On 2026-08-07 "probably" was wrong by an entire verdict. The engine on a
//   44-hour-old snapshot reported organic in BREACH at index 26. The same
//   engine, minutes later, on the live snapshot: HEALTHY at 79. Twelve orders
//   had landed in between and nothing on any page said which one you were
//   looking at.
//
// THREE STATES, AND THEY ARE NOT INTERCHANGEABLE
//
//   live         it answered and its newest data is recent enough to use
//   stale        it answered and its data has stopped moving
//   unreachable  it did not answer, or was never configured
//
//   `stale` means a person stopped filling something in. `unreachable` means a
//   credential or a service. Collapsing them into "not ok" sends whoever picks
//   it up to the wrong place, so they stay distinct and each row carries the
//   specific thing to do about it.
//
// WHERE THE ROWS COME FROM
//
//   Two halves, joined here. The engine checks what it owns (the snapshot
//   endpoint, the gig capture, the tracker) and writes the result into the run
//   JSON. The hub probes what it owns (the two Supabase stores) live on every
//   page load, because a credential should live in exactly one place and a
//   status page is not a good enough reason to copy one.
//
//   So the engine half is as fresh as the last run and says so, and the store
//   half is as fresh as this request.
//
// READ ONLY. Nothing on this page writes anywhere.

import { isMissing } from '../lib/data.js';

import {
  html,
  join,
  safe,
  missing,
  num,
  glyph,
  pill,
  panelHead,
  banner,
  why,
  info,
  empty,
  dateShort,
  dateTextOnly,
  statCard,
  statGrid,
} from './layout.js';

const TONE = { live: 'ok', stale: 'warn', unreachable: 'crit' };
const WORD = { live: 'Live', stale: 'Stale', unreachable: 'Unreachable' };

/** Worst-first. A page that leads with the healthy feeds is a page on which a
 *  broken feed stays broken. */
const RANK = { unreachable: 0, stale: 1, live: 2 };

function ageLabel(hours) {
  if (hours === null || hours === undefined) return missing();
  if (hours < 1) return html`under an hour`;
  if (hours < 48) return html`${num(Math.round(hours))} hours`;
  return html`${num(Math.round(hours / 24))} days`;
}

export function render(ctx) {
  const d = ctx.data || {};
  const engine = Array.isArray(d.engineFeeds) ? d.engineFeeds : null;
  const stores = Array.isArray(d.storeFeeds) ? d.storeFeeds : null;
  const runDate = d.runDate || null;

  // ---- neither half could be read -------------------------------------------
  if (engine === null && stores === null) {
    return {
      title: 'The feed check itself could not run',
      kicker: 'Feeds',
      ticker: [],
      html: html`${banner(
        'crit',
        'Neither the last run nor the live stores could be read.',
        html`This page is the one that is supposed to tell you when something is wrong, so it failing is
          its own kind of answer: treat every figure elsewhere in the hub as unverified until this comes
          back.`
      )}`,
    };
  }

  const rows = [...(engine || []), ...(stores || [])].sort(
    (a, b) => RANK[a.status] - RANK[b.status] || a.label.localeCompare(b.label)
  );

  const counts = { live: 0, stale: 0, unreachable: 0 };
  for (const r of rows) counts[r.status] = (counts[r.status] || 0) + 1;

  // A HALF THAT COULD NOT BE READ IS NOT A CLEAN BILL OF HEALTH.
  //
  // The verdict was taken from the rows that arrived, so with the engine half
  // unreadable the page reported "All 2 feeds live" while four feeds were
  // simply absent from the table. That is the one direction this page must
  // never be wrong in: it converts "I should go and check" into "it said it
  // was fine". Caught by its own test before it shipped.
  const halfMissing = engine === null || stores === null;
  const worst = halfMissing ? 'unreachable' : rows.length ? rows[0].status : 'unreachable';

  // ---- the verdict ----------------------------------------------------------

  const verdictBanner = halfMissing
    ? banner(
        'crit',
        html`${engine === null && stores === null ? 'Neither half' : 'One half'} of the feed check could not be read.`,
        html`${num(rows.length)} feed${rows.length === 1 ? '' : 's'} reported and the rest are unknown, so
          this page cannot tell you the system is healthy. Unknown is not the same as fine.`
      )
    : worst === 'live'
      ? banner(
          'info',
          html`All ${num(rows.length)} feeds are current.`,
          html`Every figure in this hub is built on data that is moving. This is the state to expect and it
            is worth nothing unless the other two states are loud, which is why this page leads with the
            worst feed rather than with this line.`
        )
      : banner(
          worst === 'unreachable' ? 'crit' : 'warn',
          worst === 'unreachable'
            ? html`${num(counts.unreachable)} feed${counts.unreachable === 1 ? '' : 's'} cannot be reached.`
            : html`${num(counts.stale)} feed${counts.stale === 1 ? '' : 's'} ${counts.stale === 1 ? 'has' : 'have'} stopped moving.`,
          html`Anything computed from ${counts.unreachable + counts.stale === 1 ? 'it' : 'them'} describes
            the last day ${counts.unreachable + counts.stale === 1 ? 'it' : 'they'} had data for, whatever
            date is printed on the page. Each row below says what to do.`
        );

  // ---- the table ------------------------------------------------------------

  const table = html`<section class="panel">
      ${panelHead(
        html`Every feed, ${num(rows.length)}`,
        worst === 'live' ? 'live' : 'missing',
        runDate ? `Engine half as of the ${dateTextOnly(runDate)} run · stores probed just now` : 'Stores probed just now',
        'Worst first. A page that leads with the healthy feeds is a page on which a broken feed stays broken.'
      )}
      ${!rows.length
        ? empty('No feed reported at all.')
        : html`<div class="tablewrap">
            <table class="table table--wide">
              <thead>
                <tr><th>Feed</th><th>State</th><th>Newest data</th><th class="r">Age</th><th>Where it comes from</th></tr>
              </thead>
              <tbody>
                ${join(
                  rows.map(
                    (r) => html`<tr class="${safe(r.status === 'live' ? '' : r.status === 'stale' ? 'row-attn' : 'row-late')}">
                      <td><span class="cell-name">${r.label}</span></td>
                      <td>${pill(TONE[r.status] || 'info', WORD[r.status] || r.status)}</td>
                      <td>${r.as_of ? dateShort(r.as_of) : missing()}</td>
                      <td class="r cell-figure">${ageLabel(r.age_hours)}</td>
                      <td><span class="caption">${r.source}</span></td>
                    </tr>`
                  )
                )}
              </tbody>
            </table>
            <p class="tablehint" aria-hidden="true">Scroll sideways for more columns →</p>
          </div>`}
    </section>`;

  // ---- what to do about each one that is not live ---------------------------

  const broken = rows.filter((r) => r.status !== 'live');

  const actions = html`<section class="panel">
      ${panelHead(
        html`Needs attention, ${num(broken.length)}`,
        broken.length ? 'missing' : 'live',
        broken.length ? 'One entry per feed that is not live' : 'Nothing to do',
        'Each one names the specific thing that would fix it, because "the data is old" is a symptom and the fixes are all different.'
      )}
      ${!broken.length
        ? empty('Every feed is current. Nothing here needs a person.')
        : html`<ul class="plainlist">
            ${join(
              broken.map(
                (r) => html`<li>
                  ${pill(TONE[r.status], WORD[r.status])}
                  <strong>${r.label}</strong>
                  <p class="caption">${r.detail}</p>
                  ${r.fix ? html`<p>${r.fix}</p>` : ''}
                </li>`
              )
            )}
          </ul>`}
      ${why(
        'Why stale and unreachable are two words and not one',
        html`<p>
            <em>Stale</em> means the feed answered and its data has stopped moving: somebody has stopped
            filling something in, and the fix is a person. <em>Unreachable</em> means it did not answer at
            all, or was never configured: the fix is a credential or a service.
          </p>
          <p>
            Both render as "the number is wrong", which is why it is tempting to have one word for them.
            They send you to completely different places, so they get two.
          </p>`
      )}
    </section>`;

  // ---- the split ------------------------------------------------------------

  const provenance = html`<section class="panel">
      ${panelHead('Who checks what', 'live', 'By design',
        'Each side probes only what it already holds the credentials for. A status page is not a good enough reason to copy a key into a second place.')}
      ${statGrid([
        statCard('Checked by the engine', num((engine || []).length), {
          sub: runDate ? `as of the ${dateTextOnly(runDate)} run` : 'no run found',
          tone: engine === null ? 'warn' : '',
        }),
        statCard('Probed by this page', num((stores || []).length), {
          sub: 'live, on every load',
          tone: stores === null ? 'warn' : '',
        }),
        statCard('Live', num(counts.live), { tone: 'ok' }),
        statCard('Not live', num(counts.stale + counts.unreachable), {
          tone: counts.stale + counts.unreachable ? 'warn' : '',
        }),
      ])}
      ${engine === null
        ? html`<p class="note note--neg">
            ${glyph('crit')} The last run could not be read, so the sheet, gig and tracker feeds are not
            represented above at all. They are not healthy and they are not broken; they are unknown.
          </p>`
        : ''}
      ${stores === null
        ? html`<p class="note note--neg">
            ${glyph('crit')} The two Supabase stores could not be probed, so the shift logger and the
            impressions board are missing from the table rather than shown as failing.
          </p>`
        : ''}
    </section>`;

  const title = halfMissing
    ? `Feed check incomplete, ${rows.length} of the feeds reported`
    : worst === 'live'
      ? `All ${rows.length} feeds live`
      : worst === 'unreachable'
        ? `${counts.unreachable} feed${counts.unreachable === 1 ? '' : 's'} unreachable, ${counts.stale} stale`
        : `${counts.stale} feed${counts.stale === 1 ? '' : 's'} stale`;

  return {
    title,
    kicker: 'Feeds · every input, and whether it is moving',
    ticker: [
      { label: 'Feeds', value: num(rows.length) },
      { label: 'Live', value: num(counts.live) },
      { label: 'Stale', value: num(counts.stale), sub: counts.stale ? 'data stopped moving' : '' },
      { label: 'Unreachable', value: num(counts.unreachable), sub: counts.unreachable ? 'key or service' : '' },
    ],
    html: html`${verdictBanner}
      ${table}
      ${actions}
      ${provenance}
      <div class="provenance-bar">
        <span>The engine half is written into the run JSON on each run and is therefore as fresh as the
          last run. The store half is probed on every load of this page.</span>
        <span>Ages are measured against the newest row each feed holds for this profile, not against when
          it was last touched.</span>
        <span>Nothing on this page writes anywhere, and no key is read by anything that does not already
          hold it.</span>
      </div>`,
  };
}

export default render;
